docs/ai/how-to/snippets/semantic-kernel/LocalModelExamples.cs

```diff
@@ -0,0 +1,127 @@
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.SemanticKernel;
+using Microsoft.SemanticKernel.ChatCompletion;
+using Microsoft.SemanticKernel.TextGeneration;
+
+class LocalModelExamples
+{
+    public static async Task Examples()
+    {
+        AddTextGenerationServiceExample();
+        AddChatCompletionServiceExample();
+        await UseTextGenerationServiceExample();
+        await UseChatCompletionServiceExample();
+    }
+
+    static Kernel AddTextGenerationServiceExample()
+    {
+        // <addTextService>
+        IKernelBuilder builder = Kernel.CreateBuilder();
+
+        // Add your text generation service as a singleton instance
+        var service = new MyTextGenerationService();
+        builder.Services.AddKeyedSingleton<ITextGenerationService>("myTextService1", service);
+
+        // Alternatively, add your text generation service as a factory method
+        builder.Services.AddKeyedSingleton<ITextGenerationService>("myTextService2", (_, _) => new MyTextGenerationService());
+
+        // Add any other Kernel services or configurations
+        // ...
+        Kernel kernel = builder.Build();
+        // </addTextService>
+
+        return builder.Build();
```

Comment:
Hmm. Is this idempotent? Maybe this instead:

```suggestion
        return kernel;
```